### PR TITLE
Fix time map within tolerance check and raise python exception instead of util_abort

### LINKED ISSUE #4611: Regression: __conf_instance_get_restart_nr
Running: ert gui drogon_dst.ert on the latest tag has a regression to previous versions:

Error message: __conf_instance_get_restart_nr: Failed to look up restart nr correctly

### PR DESCRIPTION
**Pull request by oyvindeide**
**Issue**
Resolves #4611 

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/clib/lib/enkf/obs_vector.cpp [hunk 1]
```diff
@@ -53,6 +54,21 @@ struct obs_vector_struct {
     std::vector<int> step_list;
 };
 
+static int find_within_tolerance(std::vector<time_t> time_map, time_t external_time) {
+    int nearest_index = -1;
+    double nearest_diff = 9999999.9;
```
Comment (pinkwah) [inline src/clib/lib/enkf/obs_vector.cpp:59]: Use infinity to show intent. Having just some arbitrary number of nines might convey the same, but it is silly.

```suggestion
    double nearest_diff = INFINITY;
```

or the more C++ (and therefore more wordy):
```suggestion
    auto nearest_diff = std::numeric_limits<double>::infinity();
```
(Might need to `#include <limits>`)

**Comment by oyvindeide** [comment]
test libres please

File: src/clib/lib/enkf/enkf_obs.cpp [hunk 2]
```diff
@@ -212,8 +212,8 @@ void enkf_obs_add_obs_vector(enkf_obs_type *enkf_obs,
     if (vector != NULL) {
         const char *obs_key = obs_vector_get_key(vector);
         if (hash_has_key(enkf_obs->obs_hash, obs_key))
-            util_abort("%s: Observation with key:%s already added.\n", __func__,
-                       obs_key);
+            throw std::runtime_error(
+                fmt::format("Duplicate observation: {}", obs_key));
```
Comment (pinkwah) [inline src/clib/lib/enkf/enkf_obs.cpp:216]: We provide a convenience exception class for this:

```suggestion
            throw exc::runtime_error("Duplicate observation: {}", obs_key);
```

I think it's in `#include <ert/except.hpp>` if not already included.

File: src/clib/lib/enkf/obs_vector.cpp [hunk 3]
```diff
@@ -63,9 +80,10 @@ __conf_instance_get_restart_nr(const conf_instance_type *conf_instance,
         obs_restart_nr =
             conf_instance_get_item_value_int(conf_instance, "RESTART");
         if (obs_restart_nr >= time_map.size())
-            util_abort("%s: Observation %s occurs at restart %i, but history "
-                       "file has only %i restarts.\n",
-                       __func__, obs_key, obs_restart_nr, time_map.size());
+            throw std::out_of_range(
+                fmt::format("Observation: {} at restart {}, but history "
+                            "file has only {} restarts.",
+                            obs_key, obs_restart_nr, time_map.size()));
```
Comment (pinkwah) [inline src/clib/lib/enkf/obs_vector.cpp:86]: Ditto. We have `exc::out_of_range`

File: src/clib/lib/enkf/obs_vector.cpp [hunk 4]
```diff
@@ -81,29 +99,22 @@ __conf_instance_get_restart_nr(const conf_instance_type *conf_instance,
             conf_instance_get_item_value_double(conf_instance, "HOURS");
         util_inplace_forward_seconds_utc(&obs_time, hours * 3600);
     } else
-        util_abort("%s: Internal error. Invalid conf_instance?\n", __func__);
-
-    auto it = std::lower_bound(time_map.begin(), time_map.end(), obs_time);
-    const time_t tolerance = 30;
-    if (it == time_map.end() || std::abs(obs_time - *it) >= tolerance) {
+        throw std::logic_error(fmt::format(
+            "Observation: {} missing DATE, DAYS or HOURS", obs_key));
```
Comment (pinkwah) [inline src/clib/lib/enkf/obs_vector.cpp:103]: `logic_error` isn't in `except.hpp` but can be added (one-line addition to that file)

File: src/clib/lib/enkf/obs_vector.cpp [hunk 5]
```diff
@@ -81,29 +99,22 @@ __conf_instance_get_restart_nr(const conf_instance_type *conf_instance,
...
+    obs_restart_nr = find_within_tolerance(time_map, obs_time);
+    if (obs_restart_nr < 0) {
+        std::string error = fmt::format(
```
Comment (pinkwah) [inline src/clib/lib/enkf/obs_vector.cpp:106]: I personally prefer to not split up exception message strings unless absolutely necessary. But it's personal taste so up to you.

Comment (oyvindeide) [reply src/clib/lib/enkf/obs_vector.cpp:106]: It was just a lot of repretition if this was not done, and a blanket if someone is using `HOURS`, which may or may not be possible

File: src/clib/lib/enkf/obs_vector.cpp [hunk 6]
```diff
@@ -53,6 +54,22 @@ struct obs_vector_struct {
     std::vector<int> step_list;
 };
 
+static int find_within_tolerance(std::vector<time_t> time_map,
```
Comment (dafeda) [inline src/clib/lib/enkf/obs_vector.cpp:57]: Perhaps change name to `find_within_tolerance` or something like that.

File: src/clib/lib/enkf/obs_vector.cpp [hunk 7]
```diff
@@ -53,6 +54,22 @@ struct obs_vector_struct {
...
+                                 time_t external_time) {
+    int nearest_index = -1;
+    auto nearest_diff = std::numeric_limits<double>::infinity();
+    for (auto &&[index, time] : iter::enumerate(time_map)) {
+        time_t diff = std::abs(external_time - time);
+        if (diff == 0)
+            return index;
+        if (diff < 30 && diff < nearest_diff) {
```
Comment (dafeda) [inline src/clib/lib/enkf/obs_vector.cpp:65]: Why is the tolerance set to 30 seconds?

File: src/clib/lib/enkf/obs_vector.cpp [hunk 8]
```diff
@@ -81,29 +99,22 @@ __conf_instance_get_restart_nr(const conf_instance_type *conf_instance,
...
-
-    auto it = std::lower_bound(time_map.begin(), time_map.end(), obs_time);
-    const time_t tolerance = 30;
-    if (it == time_map.end() || std::abs(obs_time - *it) >= tolerance) {
+        throw exc::runtime_error("Observation: {} missing DATE, DAYS or HOURS",
+                                 obs_key);
+    obs_restart_nr = find_within_tolerance(time_map, obs_time);
+    if (obs_restart_nr < 0) {
+        std::string error = fmt::format(
+            "Observation: {} failed to look up report step", obs_key);
```
Comment (dafeda) [inline src/clib/lib/enkf/obs_vector.cpp:107]: Do we need the reference to `report_step` here?

File: src/clib/lib/enkf/obs_vector.cpp (hunk 7 above)
Comment (oyvindeide) [reply src/clib/lib/enkf/obs_vector.cpp:65]: No good reason, it was always like that. It was input to the original function:
```c++
int time_map_lookup_time_with_tolerance(time_map_type *map, time_t time,
                                        int seconds_before_tolerance,
                                        int seconds_after_tolerance) {
    int nearest_index = -1;
    pthread_rwlock_rdlock(&map->rw_lock);
    {
        if (time_map_valid_time__(map, time)) {
            time_t nearest_diff = 999999999999;
            int current_index = 0;
            while (true) {
                time_t diff = time - time_map_iget__(map, current_index);
                if (diff == 0) {
                    nearest_index = current_index;
                    break;
                }

                if (std::fabs(diff) < nearest_diff) {
                    bool inside_tolerance = true;
                    if (seconds_after_tolerance >= 0) {
                        if (diff >= seconds_after_tolerance)
                            inside_tolerance = false;
                    }

                    if (seconds_before_tolerance >= 0) {
                        if (diff <= -seconds_before_tolerance)
                            inside_tolerance = false;
                    }

                    if (inside_tolerance) {
                        nearest_diff = diff;
                        nearest_index = current_index;
                    }
                }

                current_index++;

                if (current_index >= time_t_vector_size(map->map))
                    break;
            }
        }
    }
    pthread_rwlock_unlock(&map->rw_lock);
    return nearest_index;
```
but it was hard coded to 30 seconds

File: src/clib/lib/enkf/obs_vector.cpp (hunk 8 above)
Comment (oyvindeide) [reply src/clib/lib/enkf/obs_vector.cpp:107]: Just say: `failed to match time for..` or something instead?

File: src/clib/lib/enkf/obs_vector.cpp (hunk 6 above)
Comment (oyvindeide) [reply src/clib/lib/enkf/obs_vector.cpp]: Changed the name, sorry for force pushing

File: src/clib/lib/enkf/obs_vector.cpp (hunk 7 above)
Comment (dafeda) [reply src/clib/lib/enkf/obs_vector.cpp:65]: Add a comment along the lines of "The tolerance of 30s is arbitrarily chosen."?

File: tests/unit_tests/c_wrappers/res/enkf/test_summary_observation.py [hunk 9]
```diff
@@ -16,3 +24,76 @@ def test_std_scaling():
...
+
+
+@pytest.mark.parametrize(
+    "time_delta, expectation",
+    [
+        pytest.param(
+            "1.000347222", does_not_raise(), id="30 seconds offset from 1 day"
+        ),
+        pytest.param(
+            "0.999664355", does_not_raise(), id="~30 seconds offset from 1 day"
```
Comment (dafeda) [inline tests/unit_tests/c_wrappers/res/enkf/test_summary_observation.py:47]: Should the id here be -30 seconds and not ~30 seconds?

File: src/clib/lib/enkf/obs_vector.cpp (hunk 8 above)
Comment (dafeda) [reply src/clib/lib/enkf/obs_vector.cpp:107]: Sounds good 👍

File: tests/unit_tests/c_wrappers/res/enkf/test_summary_observation.py (hunk 9 above)
Comment (oyvindeide) [reply tests/unit_tests/c_wrappers/res/enkf/test_summary_observation.py:47]: It was intentionally set to `~`, as it is not exactly -30 seconds

Comment (dafeda) [reply tests/unit_tests/c_wrappers/res/enkf/test_summary_observation.py:47]: Ah I see, thanks.

**Review by dafeda — APPROVED** [review]
LGTM 👍

**Pull request merged**